File: frontend/netlify/functions/records.py
"""
Netlify Serverless Function: Record Management
Endpoint: /.netlify/functions/records
Methods: GET, POST, PUT, DELETE
"""

import json
import logging
import sys
import os
import re

# Add shared modules to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'shared'))

from db_utils import (
    get_db_connection,
    put_db_connection,
    create_record,
    fetch_records_by_encounter,
    fetch_record_by_id,
    update_record,
    delete_record,
    DatabaseError,
    RecordNotFoundError
)
from pydantic import BaseModel, Field, ValidationError

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class RecordCreateData(BaseModel):
    """Pydantic model for record creation"""
    encounter_id: int = Field(..., gt=0)
    record_date: str = Field(..., pattern=r"^\d{4}-\d{2}-\d{2}$")
    details: str = Field(..., min_length=1)


def handler(event, context):
    """Netlify serverless function handler for record management."""
    headers = {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
    }

    if event.get('httpMethod') == 'OPTIONS':
        return {'statusCode': 200, 'headers': headers, 'body': ''}

    conn = None
    try:
        conn = get_db_connection()
        if not conn:
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({'error': 'Database connection failed'})
            }

        method = event.get('httpMethod')
        path = event.get('path', '')

        # Extract IDs
        record_match = re.search(r'/records/(\d+)', path)
        encounter_match = re.search(r'/encounters/(\d+)/records', path)

        record_id = int(record_match.group(1)) if record_match else None
        encounter_id = int(encounter_match.group(1)) if encounter_match else None

        if method == 'GET':
            if record_id:
                return get_record(conn, record_id, headers)
            elif encounter_id:
                return get_encounter_records(conn, encounter_id, headers)
            return {'statusCode': 400, 'headers': headers, 'body': json.dumps({'error': 'Missing ID'})}

        elif method == 'POST' and encounter_id:
            return create_new_record(conn, encounter_id, event, headers)

        elif method == 'PUT' and record_id:
            return update_existing_record(conn, record_id, event, headers)

        elif method == 'DELETE' and record_id:
            return delete_existing_record(conn, record_id, headers)

        return {'statusCode': 400, 'headers': headers, 'body': json.dumps({'error': 'Invalid request'})}

    except Exception as e:
        logger.exception(f"Error: {e}")
        if conn:
            conn.rollback()
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}
    finally:
        if conn:
            put_db_connection(conn)


def get_encounter_records(conn, encounter_id, headers):
    """Get all records for an encounter"""
    try:
        records = fetch_records_by_encounter(conn=conn, encounter_id=encounter_id)
        return {'statusCode': 200 if records else 404, 'headers': headers,
                'body': json.dumps(records if records else {'error': 'No records found'})}
    except Exception as e:
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}


def get_record(conn, record_id, headers):
    """Get a single record"""
    try:
        record = fetch_record_by_id(conn=conn, record_id=record_id)
        if record:
            return {'statusCode': 200, 'headers': headers, 'body': json.dumps(record)}
        return {'statusCode': 404, 'headers': headers, 'body': json.dumps({'error': 'Not found'})}
    except Exception as e:
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}


def create_new_record(conn, encounter_id, event, headers):
    """Create a new record"""
    try:
        body = json.loads(event['body']) if isinstance(event.get('body'), str) else event.get('body', {})
        validated_data = RecordCreateData.model_validate({**body, "encounter_id": encounter_id})

        record_id = create_record(
            conn=conn,
            encounter_id=validated_data.encounter_id,
            record_date=validated_data.record_date,
            details=validated_data.details
        )

        if record_id:
            new_record = fetch_record_by_id(conn=conn, record_id=record_id)
            return {'statusCode': 201, 'headers': headers,
                    'body': json.dumps({'message': 'Record created', 'record': new_record})}
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': 'Failed'})}
    except ValidationError as e:
        return {'statusCode': 400, 'headers': headers, 'body': json.dumps({'error': e.errors()})}
    except Exception as e:
        conn.rollback()
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}


def update_existing_record(conn, record_id, event, headers):
    """Update a record"""
    try:
        body = json.loads(event['body']) if isinstance(event.get('body'), str) else event.get('body', {})
        updated = update_record(
            conn=conn,
            record_id=record_id,
            record_date=body.get('record_date'),
            details=body.get('details'),
            user_id=body.get('user_id')
        )

        if updated > 0:
            record = fetch_record_by_id(conn=conn, record_id=record_id)
            return {'statusCode': 200, 'headers': headers,
                    'body': json.dumps({'message': 'Updated', 'record': record})}
        return {'statusCode': 404, 'headers': headers, 'body': json.dumps({'error': 'Not found'})}
    except Exception as e:
        conn.rollback()
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}


def delete_existing_record(conn, record_id, headers):
    """Delete a record"""
    try:
        deleted = delete_record(conn=conn, record_id=record_id)
        if deleted > 0:
            return {'statusCode': 200, 'headers': headers, 'body': json.dumps({'message': 'Deleted'})}
        return {'statusCode': 404, 'headers': headers, 'body': json.dumps({'error': 'Not found'})}
    except Exception as e:
        conn.rollback()
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}

File: frontend/netlify/functions/records.js
const { Pool } = require('pg');

const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
  ssl: process.env.NODE_ENV === 'production' ? { rejectUnauthorized: false } : false
});

const headers = {
  'Content-Type': 'application/json',
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type'
};

exports.handler = async (event) => {
  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers, body: '' };
  }

  const recordMatch = event.path.match(/\/records\/(\d+)/);
  const encounterMatch = event.path.match(/\/encounters\/(\d+)\/records/);

  const recordId = recordMatch ? parseInt(recordMatch[1]) : null;
  const encounterId = encounterMatch ? parseInt(encounterMatch[1]) : null;

  try {
    switch (event.httpMethod) {
      case 'GET':
        if (recordId) return await getRecord(recordId);
        if (encounterId) return await getEncounterRecords(encounterId);
        return { statusCode: 400, headers, body: JSON.stringify({ error: 'Encounter ID or Record ID required' }) };
      case 'POST':
        if (!encounterId) return { statusCode: 400, headers, body: JSON.stringify({ error: 'Encounter ID required' }) };
        return await createRecord(encounterId, JSON.parse(event.body || '{}'));
      case 'PUT':
        if (!recordId) return { statusCode: 400, headers, body: JSON.stringify({ error: 'Record ID required' }) };
        return await updateRecord(recordId, JSON.parse(event.body || '{}'));
      case 'DELETE':
        if (!recordId) return { statusCode: 400, headers, body: JSON.stringify({ error: 'Record ID required' }) };
        return await deleteRecord(recordId);
      default:
        return { statusCode: 405, headers, body: JSON.stringify({ error: 'Method not allowed' }) };
    }
  } catch (error) {
    console.error('Record operation error:', error);
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: 'Internal server error', details: error.message })
    };
  }
};

async function getEncounterRecords(encounterId) {
  const result = await pool.query(
    'SELECT * FROM records WHERE encounter_id = $1 ORDER BY record_date DESC',
    [encounterId]
  );
  return { statusCode: 200, headers, body: JSON.stringify(result.rows) };
}

async function getRecord(id) {
  const result = await pool.query('SELECT * FROM records WHERE record_id = $1', [id]);
  if (result.rows.length === 0) {
    return { statusCode: 404, headers, body: JSON.stringify({ error: 'Record not found' }) };
  }
  return { statusCode: 200, headers, body: JSON.stringify(result.rows[0]) };
}

async function createRecord(encounterId, data) {
  const result = await pool.query(
    `INSERT INTO records (encounter_id, record_date, details, user_id)
     VALUES ($1, $2, $3, $4) RETURNING *`,
    [encounterId, data.record_date, data.details, data.user_id]
  );
  return {
    statusCode: 201,
    headers,
    body: JSON.stringify({ message: 'Record created successfully', record: result.rows[0] })
  };
}

async function updateRecord(id, data) {
  const fields = [];
  const values = [];
  let idx = 1;

  if (data.record_date) { fields.push(`record_date = $${idx++}`); values.push(data.record_date); }
  if (data.details) { fields.push(`details = $${idx++}`); values.push(data.details); }

  if (fields.length === 0) {
    return { statusCode: 400, headers, body: JSON.stringify({ error: 'No fields to update' }) };
  }

  fields.push(`updated_at = NOW()`);
  values.push(id);

  const result = await pool.query(
    `UPDATE records SET ${fields.join(', ')} WHERE record_id = $${idx} RETURNING *`,
    values
  );

  if (result.rows.length === 0) {
    return { statusCode: 404, headers, body: JSON.stringify({ error: 'Record not found' }) };
  }
  return {
    statusCode: 200,
    headers,
    body: JSON.stringify({ message: 'Record updated successfully', record: result.rows[0] })
  };
}

async function deleteRecord(id) {
  const result = await pool.query('DELETE FROM records WHERE record_id = $1 RETURNING *', [id]);
  if (result.rows.length === 0) {
    return { statusCode: 404, headers, body: JSON.stringify({ error: 'Record not found' }) };
  }
  return { statusCode: 200, headers, body: JSON.stringify({ message: 'Record deleted successfully' }) };
}
